File: cliphist_picker/window.py
import threading

import gi

gi.require_version("Gtk", "4.0")
gi.require_version("Gtk4LayerShell", "1.0")

from gi.repository import Gdk, GLib, Gtk, Gtk4LayerShell

from .clipboard import (
    copy_image_to_clipboard,
    copy_text_to_clipboard,
    decode_entry,
    paste_via_wtype,
)
from .config import Config
from .models import ClipboardEntry, EntryType
from .row import create_row, load_preview_from_bytes, load_thumbnail_from_bytes


class PickerWindow(Gtk.Window):
    def __init__(self, app: Gtk.Application, entries: list[ClipboardEntry], config: Config):
        super().__init__(application=app)
        self.app = app
        self.config = config
        self.entries = entries
        self._image_cache: dict[str, bytes] = {}
        self._hovered_row: Gtk.ListBoxRow | None = None

        # Layer Shell setup
        Gtk4LayerShell.init_for_window(self)
        Gtk4LayerShell.set_layer(self, Gtk4LayerShell.Layer.OVERLAY)
        Gtk4LayerShell.set_keyboard_mode(
            self, Gtk4LayerShell.KeyboardMode.EXCLUSIVE,
        )
        Gtk4LayerShell.set_namespace(self, "cliphist-picker")
        Gtk4LayerShell.set_exclusive_zone(self, -1)

        # Anchor all edges -> full screen
        for edge in (
            Gtk4LayerShell.Edge.TOP,
            Gtk4LayerShell.Edge.BOTTOM,
            Gtk4LayerShell.Edge.LEFT,
            Gtk4LayerShell.Edge.RIGHT,
        ):
            Gtk4LayerShell.set_anchor(self, edge, True)

        self.remove_css_class("background")
        self.add_css_class("backdrop")

        self._build_ui()
        self._setup_input()
        self._populate()
        self._load_thumbnails_async()

        # Position preview after the window gets its allocation
        self.connect("realize", self._on_realize)

    # ── Layout ───────────────────────────────────────────────────────

    def _build_ui(self):
        # Use Overlay so the preview floats without shifting the picker
        overlay = Gtk.Overlay()

        # Full-screen backdrop — ensures the Wayland input region covers the
        # entire surface so clicks outside the picker and keyboard events
        # are forwarded by the compositor.
        backdrop = Gtk.Box()
        backdrop.set_hexpand(True)
        backdrop.set_vexpand(True)
        overlay.set_child(backdrop)

        # Picker frame (centered, overlay child)
        self.picker_frame = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)
        self.picker_frame.add_css_class("picker-frame")
        self.picker_frame.set_size_request(self.config.width, -1)
        self.picker_frame.set_halign(Gtk.Align.CENTER)
        self.picker_frame.set_valign(Gtk.Align.CENTER)

        self.scrolled = Gtk.ScrolledWindow()
        self.scrolled.set_policy(Gtk.PolicyType.NEVER, Gtk.PolicyType.AUTOMATIC)
        self.scrolled.set_max_content_height(self.config.max_height)
        self.scrolled.set_propagate_natural_height(True)
        self.scrolled.add_css_class("picker-scroll")

        self.listbox = Gtk.ListBox()
        self.listbox.set_selection_mode(Gtk.SelectionMode.SINGLE)
        self.listbox.add_css_class("picker-list")
        self.scrolled.set_child(self.listbox)
        self.picker_frame.append(self.scrolled)
        overlay.add_overlay(self.picker_frame)

        # Preview — floats to the right of center, doesn't affect picker position
        self.preview_picture = Gtk.Picture()
        self.preview_picture.set_can_shrink(True)
        self.preview_picture.set_content_fit(Gtk.ContentFit.CONTAIN)
        self.preview_picture.add_css_class("preview-image")
        self.preview_picture.set_size_request(self.config.preview_width, -1)
        self.preview_picture.set_halign(Gtk.Align.END)
        self.preview_picture.set_valign(Gtk.Align.CENTER)
        self.preview_picture.set_margin_end(
            max(20, (1920 - self.config.width) // 2 - self.config.preview_width - 40),
        )
        self.preview_picture.set_visible(False)
        overlay.add_overlay(self.preview_picture)

        self.set_child(overlay)

    def _on_realize(self, widget):
        """Position the preview based on actual screen width."""
        surface = self.get_surface()
        if surface:
            monitor = self.get_display().get_monitor_at_surface(surface)
            if monitor:
                screen_w = monitor.get_geometry().width
                # Place preview in the gap between center picker and screen edge
                gap = (screen_w - self.config.width) // 2
                margin = max(20, gap - self.config.preview_width - 24)
                self.preview_picture.set_margin_end(margin)

    # ── Input ────────────────────────────────────────────────────────

    def _setup_input(self):
        key_ctrl = Gtk.EventControllerKey()
        key_ctrl.set_propagation_phase(Gtk.PropagationPhase.CAPTURE)
        key_ctrl.connect("key-pressed", self._on_key_pressed)
        self.add_controller(key_ctrl)

        self.listbox.connect("row-activated", self._on_row_activated)
        self.listbox.connect("selected-rows-changed", self._on_selection_changed)

        # Backdrop click
        click_ctrl = Gtk.GestureClick()
        click_ctrl.connect("pressed", self._on_backdrop_click)
        self.add_controller(click_ctrl)

        # Hover for preview
        motion_ctrl = Gtk.EventControllerMotion()
        motion_ctrl.connect("motion", self._on_listbox_motion)
        motion_ctrl.connect("leave", self._on_listbox_leave)
        self.listbox.add_controller(motion_ctrl)

    def _populate(self):
        for entry in self.entries:
            row = create_row(entry, self.config)
            self.listbox.append(row)
        first = self.listbox.get_row_at_index(0)
        if first:
            self.listbox.select_row(first)
            first.grab_focus()

    # ── Keyboard ─────────────────────────────────────────────────────

    def _on_key_pressed(self, controller, keyval, keycode, state):
        name = Gdk.keyval_name(keyval)
        print(f"[DEBUG] key-pressed: {name}", flush=True)

        if name == "Escape":
            self._dismiss()
            return True

        if name in ("Return", "KP_Enter"):
            selected = self.listbox.get_selected_row()
            if selected:
                self._activate_entry(selected.entry)
            return True

        if name in ("Up", "w", "W"):
            self._move_selection(-1)
            return True

        if name in ("Down", "s", "S"):
            self._move_selection(1)
            return True

        return False

    def _move_selection(self, delta: int):
        selected = self.listbox.get_selected_row()
        if selected is None:
            target_idx = 0
        else:
            target_idx = selected.get_index() + delta

        target_idx = max(0, min(target_idx, len(self.entries) - 1))
        target_row = self.listbox.get_row_at_index(target_idx)
        if target_row:
            self.listbox.select_row(target_row)
            target_row.grab_focus()

    # ── Selection / Preview ──────────────────────────────────────────

    def _on_selection_changed(self, listbox):
        selected = listbox.get_selected_row()
        if selected is None:
            self.preview_picture.set_visible(False)
            return

        entry: ClipboardEntry = selected.entry
        if entry.has_image and entry.image_raw_line:
            self._show_preview_for(entry)
        else:
            self.preview_picture.set_visible(False)

    def _on_listbox_motion(self, controller, x, y):
        row = self.listbox.get_row_at_y(y)
        if row is self._hovered_row:
            return
        self._hovered_row = row
        if row is None:
            return
        entry: ClipboardEntry = row.entry
        if entry.has_image and entry.image_raw_line:
            self._show_preview_for(entry)
        else:
            # Only hide preview if keyboard selection also has no image
            selected = self.listbox.get_selected_row()
            if selected is None or not selected.entry.has_image:
                self.preview_picture.set_visible(False)

    def _on_listbox_leave(self, controller):
        self._hovered_row = None
        # Fall back to keyboard selection preview
        selected = self.listbox.get_selected_row()
        if selected and selected.entry.has_image and selected.entry.image_raw_line:
            self._show_preview_for(selected.entry)
        else:
            self.preview_picture.set_visible(False)

    def _show_preview_for(self, entry: ClipboardEntry):
        raw_line = entry.image_raw_line
        if raw_line in self._image_cache:
            self._render_preview(self._image_cache[raw_line])
        else:
            def decode_and_show():
                data = decode_entry(raw_line)
                self._image_cache[raw_line] = data
                GLib.idle_add(self._render_preview, data)
            threading.Thread(target=decode_and_show, daemon=True).start()

    def _render_preview(self, image_bytes: bytes):
        texture = load_preview_from_bytes(
            image_bytes, self.config.preview_width, self.config.preview_max_height,
        )
        if texture:
            self.preview_picture.set_paintable(texture)
            self.preview_picture.set_visible(True)
        else:
            self.preview_picture.set_visible(False)
        return False  # for GLib.idle_add

    # ── Activation ───────────────────────────────────────────────────

    def _on_row_activated(self, listbox, row):
        self._activate_entry(row.entry)

    def _activate_entry(self, entry: ClipboardEntry):
        if entry.entry_type == EntryType.IMAGE:
            mime = f"image/{entry.image_format or 'png'}"
            copy_image_to_clipboard(entry.primary_raw_line, mime)
        else:
            copy_text_to_clipboard(entry.primary_raw_line)

        if self.config.close_on_select:
            self.set_visible(False)

        if self.config.auto_paste:
            GLib.timeout_add(self.config.paste_delay_ms, self._do_paste)
        else:
            self._dismiss()

    def _do_paste(self):
        paste_via_wtype()
        self._dismiss()
        return False

    def _dismiss(self):
        print("[DEBUG] _dismiss called", flush=True)
        self.close()
        self.app.quit()

    # ── Backdrop Click ───────────────────────────────────────────────

    def _on_backdrop_click(self, gesture, n_press, x, y):
        widget_at = self.pick(x, y, Gtk.PickFlags.DEFAULT)
        inside = self._is_inside_picker(widget_at) if widget_at else False
        print(f"[DEBUG] backdrop-click: widget={widget_at}, inside={inside}", flush=True)
        # Dismiss if click hit transparent area (None) or anything outside the picker
        if widget_at is None or not self._is_inside_picker(widget_at):
            self._dismiss()

    def _is_inside_picker(self, widget):
        """Check if a widget is a descendant of the picker frame or preview."""
        while widget is not None:
            if widget is self.picker_frame or widget is self.preview_picture:
                return True
            widget = widget.get_parent()
        return False

    # ── Async Thumbnails ─────────────────────────────────────────────

    def _load_thumbnails_async(self):
        def load_all():
            for i, entry in enumerate(self.entries):
                if not entry.has_image or not entry.image_raw_line:
                    continue
                raw_line = entry.image_raw_line
                if raw_line not in self._image_cache:
                    data = decode_entry(raw_line)
                    self._image_cache[raw_line] = data
                else:
                    data = self._image_cache[raw_line]

                if data:
                    # Create texture on main thread (GdkPixbuf not thread-safe)
                    GLib.idle_add(self._create_and_set_thumbnail, i, data)

        threading.Thread(target=load_all, daemon=True).start()

    def _create_and_set_thumbnail(self, index: int, data: bytes):
        texture = load_thumbnail_from_bytes(data, self.config.thumbnail_size)
        if texture:
            self._set_thumbnail(index, texture)
        return False

    def _set_thumbnail(self, index: int, texture):
        row = self.listbox.get_row_at_index(index)
        if row and hasattr(row, "thumbnail_widget"):
            row.thumbnail_widget.set_paintable(texture)
        return False

File: src/window.c
#include "window.h"

#include <gtk4-layer-shell.h>
#include <string.h>

#include "clipboard.h"
#include "row.h"

/* ── Forward declarations ──────────────────────────────────────── */

static void     build_ui(PickerWindow *win);
static void     setup_input(PickerWindow *win);
static void     populate(PickerWindow *win);
static void     load_thumbnails_async(PickerWindow *win);
static void     dismiss(PickerWindow *win);
static gboolean render_preview_idle(gpointer user_data);

/* ── Image cache helpers ───────────────────────────────────────── */

static ImageData *image_data_new(guint8 *data, gsize len) {
    ImageData *img = g_new(ImageData, 1);
    img->data = data;
    img->len  = len;
    return img;
}

static void image_data_free(gpointer p) {
    ImageData *img = p;
    g_free(img->data);
    g_free(img);
}

/* ── Thumbnail thread data ─────────────────────────────────────── */

typedef struct {
    PickerWindow *win;
    int           row_index;
    guint8       *data;
    gsize         len;
} ThumbnailJob;

static gboolean create_and_set_thumbnail_idle(gpointer user_data) {
    ThumbnailJob *job = user_data;
    PickerWindow *win = job->win;

    GdkTexture *texture = load_thumbnail_from_bytes(
        job->data, job->len, win->config.thumbnail_size);

    if (texture) {
        GtkListBoxRow *row = gtk_list_box_get_row_at_index(
            win->listbox, job->row_index);
        if (row) {
            GtkPicture *thumb = GTK_PICTURE(
                g_object_get_data(G_OBJECT(row), "thumbnail-widget"));
            if (thumb)
                gtk_picture_set_paintable(thumb, GDK_PAINTABLE(texture));
        }
        g_object_unref(texture);
    }

    g_free(job);
    return G_SOURCE_REMOVE;
}

static gpointer thumbnail_thread_func(gpointer data) {
    PickerWindow *win = data;

    for (int i = 0; i < win->n_entries; i++) {
        ClipboardEntry *entry = &win->entries[i];
        if (!clipboard_entry_has_image(entry) || !entry->image_raw_line)
            continue;

        g_mutex_lock(&win->cache_mutex);
        gboolean cached = g_hash_table_contains(win->image_cache,
                                                 entry->image_raw_line);
        g_mutex_unlock(&win->cache_mutex);

        ImageData *img;
        if (!cached) {
            gsize len = 0;
            guint8 *raw = clipboard_decode(entry->image_raw_line, &len);
            if (!raw || len == 0) {
                g_free(raw);
                continue;
            }
            img = image_data_new(raw, len);
            g_mutex_lock(&win->cache_mutex);
            g_hash_table_insert(win->image_cache,
                                g_strdup(entry->image_raw_line), img);
            g_mutex_unlock(&win->cache_mutex);
        } else {
            g_mutex_lock(&win->cache_mutex);
            img = g_hash_table_lookup(win->image_cache, entry->image_raw_line);
            g_mutex_unlock(&win->cache_mutex);
            if (!img) continue;
        }

        ThumbnailJob *job = g_new0(ThumbnailJob, 1);
        job->win       = win;
        job->row_index = i;
        job->data      = img->data;
        job->len       = img->len;
        g_idle_add(create_and_set_thumbnail_idle, job);
    }

    return NULL;
}

static void load_thumbnails_async(PickerWindow *win) {
    g_thread_new("thumbnail-loader", thumbnail_thread_func, win);
}

/* ── Preview ───────────────────────────────────────────────────── */

static void render_preview(PickerWindow *win, const guint8 *data, gsize len) {
    GdkTexture *texture = load_preview_from_bytes(
        data, len, win->config.preview_width, win->config.preview_max_height);
    if (texture) {
        gtk_picture_set_paintable(win->preview_picture, GDK_PAINTABLE(texture));
        gtk_widget_set_visible(GTK_WIDGET(win->preview_picture), TRUE);
        g_object_unref(texture);
    } else {
        gtk_widget_set_visible(GTK_WIDGET(win->preview_picture), FALSE);
    }
}

typedef struct {
    PickerWindow *win;
    char         *raw_line;
} PreviewDecodeJob;

static gpointer preview_decode_thread(gpointer data) {
    PreviewDecodeJob *job = data;
    PickerWindow *win = job->win;

    gsize len = 0;
    guint8 *decoded = clipboard_decode(job->raw_line, &len);

    if (decoded && len > 0) {
        ImageData *img = image_data_new(decoded, len);
        g_mutex_lock(&win->cache_mutex);
        g_hash_table_insert(win->image_cache,
                            job->raw_line, img);
        g_mutex_unlock(&win->cache_mutex);
        job->raw_line = NULL; /* ownership transferred to cache */

        ThumbnailJob *tj = g_new0(ThumbnailJob, 1);
        tj->win  = win;
        tj->data = img->data;
        tj->len  = img->len;
        tj->row_index = -1; /* signal: this is a preview, not thumbnail */
        g_idle_add(render_preview_idle, tj);
    }

    g_free(job->raw_line);
    g_free(job);
    return NULL;
}

static gboolean render_preview_idle(gpointer user_data) {
    ThumbnailJob *job = user_data;
    render_preview(job->win, job->data, job->len);
    g_free(job);
    return G_SOURCE_REMOVE;
}

static void show_preview_for(PickerWindow *win, ClipboardEntry *entry) {
    const char *raw_line = entry->image_raw_line;
    if (!raw_line) return;

    g_mutex_lock(&win->cache_mutex);
    ImageData *img = g_hash_table_lookup(win->image_cache, raw_line);
    g_mutex_unlock(&win->cache_mutex);

    if (img) {
        render_preview(win, img->data, img->len);
    } else {
        PreviewDecodeJob *job = g_new0(PreviewDecodeJob, 1);
        job->win      = win;
        job->raw_line = g_strdup(raw_line);
        g_thread_new("preview-decode", preview_decode_thread, job);
    }
}

/* ── Keyboard ──────────────────────────────────────────────────── */

static void move_selection(PickerWindow *win, int delta) {
    GtkListBoxRow *selected = gtk_list_box_get_selected_row(win->listbox);
    int target_idx;
    if (!selected) {
        target_idx = 0;
    } else {
        target_idx = gtk_list_box_row_get_index(selected) + delta;
    }

    if (target_idx < 0) target_idx = 0;
    if (target_idx >= win->n_entries) target_idx = win->n_entries - 1;

    GtkListBoxRow *target = gtk_list_box_get_row_at_index(
        win->listbox, target_idx);
    if (target) {
        gtk_list_box_select_row(win->listbox, target);
        gtk_widget_grab_focus(GTK_WIDGET(target));
    }
}

static void activate_entry(PickerWindow *win, ClipboardEntry *entry);

static gboolean on_key_pressed(GtkEventControllerKey *ctrl,
                                guint keyval, guint keycode,
                                GdkModifierType state,
                                gpointer user_data) {
    (void)ctrl; (void)keycode; (void)state;
    PickerWindow *win = user_data;
    const char *name = gdk_keyval_name(keyval);

    if (g_strcmp0(name, "Escape") == 0) {
        dismiss(win);
        return TRUE;
    }

    if (g_strcmp0(name, "Return") == 0 || g_strcmp0(name, "KP_Enter") == 0) {
        GtkListBoxRow *row = gtk_list_box_get_selected_row(win->listbox);
        if (row) {
            ClipboardEntry *entry = g_object_get_data(
                G_OBJECT(row), "clipboard-entry");
            if (entry) activate_entry(win, entry);
        }
        return TRUE;
    }

    if (g_strcmp0(name, "Up") == 0 ||
        g_strcmp0(name, "w") == 0 ||
        g_strcmp0(name, "W") == 0) {
        move_selection(win, -1);
        return TRUE;
    }

    if (g_strcmp0(name, "Down") == 0 ||
        g_strcmp0(name, "s") == 0 ||
        g_strcmp0(name, "S") == 0) {
        move_selection(win, 1);
        return TRUE;
    }

    return FALSE;
}

/* ── Selection / Hover ─────────────────────────────────────────── */

static void on_selection_changed(GtkListBox *listbox, gpointer user_data) {
    PickerWindow *win = user_data;
    GtkListBoxRow *selected = gtk_list_box_get_selected_row(listbox);

    if (!selected) {
        gtk_widget_set_visible(GTK_WIDGET(win->preview_picture), FALSE);
        return;
    }

    ClipboardEntry *entry = g_object_get_data(
        G_OBJECT(selected), "clipboard-entry");
    if (entry && clipboard_entry_has_image(entry) && entry->image_raw_line) {
        show_preview_for(win, entry);
    } else {
        gtk_widget_set_visible(GTK_WIDGET(win->preview_picture), FALSE);
    }
}

static void on_listbox_motion(GtkEventControllerMotion *ctrl,
                               double x, double y,
                               gpointer user_data) {
    (void)ctrl; (void)x;
    PickerWindow *win = user_data;
    GtkListBoxRow *row = gtk_list_box_get_row_at_y(win->listbox, y);

    if (row == win->hovered_row) return;
    win->hovered_row = row;

    if (!row) return;

    ClipboardEntry *entry = g_object_get_data(
        G_OBJECT(row), "clipboard-entry");
    if (entry && clipboard_entry_has_image(entry) && entry->image_raw_line) {
        show_preview_for(win, entry);
    } else {
        GtkListBoxRow *selected = gtk_list_box_get_selected_row(win->listbox);
        if (!selected) {
            gtk_widget_set_visible(GTK_WIDGET(win->preview_picture), FALSE);
        } else {
            ClipboardEntry *sel_entry = g_object_get_data(
                G_OBJECT(selected), "clipboard-entry");
            if (!sel_entry || !clipboard_entry_has_image(sel_entry))
                gtk_widget_set_visible(GTK_WIDGET(win->preview_picture), FALSE);
        }
    }
}

static void on_listbox_leave(GtkEventControllerMotion *ctrl,
                              gpointer user_data) {
    (void)ctrl;
    PickerWindow *win = user_data;
    win->hovered_row = NULL;

    GtkListBoxRow *selected = gtk_list_box_get_selected_row(win->listbox);
    if (selected) {
        ClipboardEntry *entry = g_object_get_data(
            G_OBJECT(selected), "clipboard-entry");
        if (entry && clipboard_entry_has_image(entry) && entry->image_raw_line) {
            show_preview_for(win, entry);
            return;
        }
    }
    gtk_widget_set_visible(GTK_WIDGET(win->preview_picture), FALSE);
}

/* ── Activation ────────────────────────────────────────────────── */

static gboolean do_paste(gpointer user_data) {
    PickerWindow *win = user_data;
    clipboard_paste_wtype();
    dismiss(win);
    return G_SOURCE_REMOVE;
}

static void activate_entry(PickerWindow *win, ClipboardEntry *entry) {
    if (entry->entry_type == ENTRY_TYPE_IMAGE) {
        char *mime = g_strdup_printf("image/%s",
            entry->image_format ? entry->image_format : "png");
        clipboard_copy_image(entry->primary_raw_line, mime);
        g_free(mime);
    } else {
        clipboard_copy_text(entry->primary_raw_line);
    }

    if (win->config.close_on_select) {
        gtk_widget_set_visible(GTK_WIDGET(win->window), FALSE);
    }

    if (win->config.auto_paste) {
        g_timeout_add(win->config.paste_delay_ms, do_paste, win);
    } else {
        dismiss(win);
    }
}

static void on_row_activated(GtkListBox *listbox, GtkListBoxRow *row,
                              gpointer user_data) {
    (void)listbox;
    PickerWindow *win = user_data;
    ClipboardEntry *entry = g_object_get_data(
        G_OBJECT(row), "clipboard-entry");
    if (entry) activate_entry(win, entry);
}

/* ── Backdrop Click ────────────────────────────────────────────── */

static gboolean is_inside_picker(PickerWindow *win, GtkWidget *widget) {
    while (widget) {
        if (widget == GTK_WIDGET(win->picker_frame) ||
            widget == GTK_WIDGET(win->preview_picture))
            return TRUE;
        widget = gtk_widget_get_parent(widget);
    }
    return FALSE;
}

static void on_backdrop_click(GtkGestureClick *gesture,
                               int n_press, double x, double y,
                               gpointer user_data) {
    (void)gesture; (void)n_press;
    PickerWindow *win = user_data;
    GtkWidget *target = gtk_widget_pick(
        GTK_WIDGET(win->window), x, y, GTK_PICK_DEFAULT);

    if (!target || !is_inside_picker(win, target))
        dismiss(win);
}

/* ── Realize ───────────────────────────────────────────────────── */

static void on_realize(GtkWidget *widget, gpointer user_data) {
    (void)widget;
    PickerWindow *win = user_data;
    GdkSurface *surface = gtk_native_get_surface(GTK_NATIVE(win->window));
    if (!surface) return;

    GdkDisplay *display = gtk_widget_get_display(GTK_WIDGET(win->window));
    GdkMonitor *monitor = gdk_display_get_monitor_at_surface(display, surface);
    if (!monitor) return;

    GdkRectangle geom;
    gdk_monitor_get_geometry(monitor, &geom);
    int screen_w = geom.width;
    int gap = (screen_w - win->config.width) / 2;
    int margin = MAX(20, gap - win->config.preview_width - 24);
    gtk_widget_set_margin_end(GTK_WIDGET(win->preview_picture), margin);
}

/* ── Dismiss ───────────────────────────────────────────────────── */

static void dismiss(PickerWindow *win) {
    gtk_window_close(win->window);
    g_application_quit(G_APPLICATION(win->app));
}

/* ── Build UI ──────────────────────────────────────────────────── */

static void build_ui(PickerWindow *win) {
    GtkOverlay *overlay = GTK_OVERLAY(gtk_overlay_new());

    /* Full-screen backdrop for input region */
    GtkBox *backdrop = GTK_BOX(gtk_box_new(GTK_ORIENTATION_VERTICAL, 0));
    gtk_widget_set_hexpand(GTK_WIDGET(backdrop), TRUE);
    gtk_widget_set_vexpand(GTK_WIDGET(backdrop), TRUE);
    gtk_overlay_set_child(overlay, GTK_WIDGET(backdrop));

    /* Picker frame */
    win->picker_frame = GTK_BOX(gtk_box_new(GTK_ORIENTATION_VERTICAL, 0));
    gtk_widget_add_css_class(GTK_WIDGET(win->picker_frame), "picker-frame");
    gtk_widget_set_size_request(GTK_WIDGET(win->picker_frame),
                                 win->config.width, -1);
    gtk_widget_set_halign(GTK_WIDGET(win->picker_frame), GTK_ALIGN_CENTER);
    gtk_widget_set_valign(GTK_WIDGET(win->picker_frame), GTK_ALIGN_CENTER);

    /* Scrolled window */
    win->scrolled = GTK_SCROLLED_WINDOW(gtk_scrolled_window_new());
    gtk_scrolled_window_set_policy(win->scrolled,
                                    GTK_POLICY_NEVER, GTK_POLICY_AUTOMATIC);
    gtk_scrolled_window_set_max_content_height(win->scrolled,
                                                 win->config.max_height);
    gtk_scrolled_window_set_propagate_natural_height(win->scrolled, TRUE);
    gtk_widget_add_css_class(GTK_WIDGET(win->scrolled), "picker-scroll");

    /* List box */
    win->listbox = GTK_LIST_BOX(gtk_list_box_new());
    gtk_list_box_set_selection_mode(win->listbox, GTK_SELECTION_SINGLE);
    gtk_widget_add_css_class(GTK_WIDGET(win->listbox), "picker-list");
    gtk_scrolled_window_set_child(win->scrolled, GTK_WIDGET(win->listbox));
    gtk_box_append(win->picker_frame, GTK_WIDGET(win->scrolled));
    gtk_overlay_add_overlay(overlay, GTK_WIDGET(win->picker_frame));

    /* Preview picture */
    win->preview_picture = GTK_PICTURE(gtk_picture_new());
    gtk_picture_set_can_shrink(win->preview_picture, TRUE);
    gtk_picture_set_content_fit(win->preview_picture, GTK_CONTENT_FIT_CONTAIN);
    gtk_widget_add_css_class(GTK_WIDGET(win->preview_picture), "preview-image");
    gtk_widget_set_size_request(GTK_WIDGET(win->preview_picture),
                                 win->config.preview_width, -1);
    gtk_widget_set_halign(GTK_WIDGET(win->preview_picture), GTK_ALIGN_END);
    gtk_widget_set_valign(GTK_WIDGET(win->preview_picture), GTK_ALIGN_CENTER);
    gtk_widget_set_margin_end(GTK_WIDGET(win->preview_picture),
        MAX(20, (1920 - win->config.width) / 2 - win->config.preview_width - 40));
    gtk_widget_set_visible(GTK_WIDGET(win->preview_picture), FALSE);
    gtk_overlay_add_overlay(overlay, GTK_WIDGET(win->preview_picture));

    gtk_window_set_child(win->window, GTK_WIDGET(overlay));
}

/* ── Input setup ───────────────────────────────────────────────── */

static void setup_input(PickerWindow *win) {
    /* Keyboard (capture phase) */
    GtkEventControllerKey *key_ctrl = GTK_EVENT_CONTROLLER_KEY(
        gtk_event_controller_key_new());
    gtk_event_controller_set_propagation_phase(
        GTK_EVENT_CONTROLLER(key_ctrl), GTK_PHASE_CAPTURE);
    g_signal_connect(key_ctrl, "key-pressed",
        G_CALLBACK(on_key_pressed), win);
    gtk_widget_add_controller(GTK_WIDGET(win->window),
        GTK_EVENT_CONTROLLER(key_ctrl));

    /* Row activation and selection */
    g_signal_connect(win->listbox, "row-activated",
        G_CALLBACK(on_row_activated), win);
    g_signal_connect(win->listbox, "selected-rows-changed",
        G_CALLBACK(on_selection_changed), win);

    /* Backdrop click */
    GtkGestureClick *click = GTK_GESTURE_CLICK(gtk_gesture_click_new());
    g_signal_connect(click, "pressed",
        G_CALLBACK(on_backdrop_click), win);
    gtk_widget_add_controller(GTK_WIDGET(win->window),
        GTK_EVENT_CONTROLLER(click));

    /* Hover for preview */
    GtkEventControllerMotion *motion = GTK_EVENT_CONTROLLER_MOTION(
        gtk_event_controller_motion_new());
    g_signal_connect(motion, "motion",
        G_CALLBACK(on_listbox_motion), win);
    g_signal_connect(motion, "leave",
        G_CALLBACK(on_listbox_leave), win);
    gtk_widget_add_controller(GTK_WIDGET(win->listbox),
        GTK_EVENT_CONTROLLER(motion));
}

/* ── Populate ──────────────────────────────────────────────────── */

static void populate(PickerWindow *win) {
    for (int i = 0; i < win->n_entries; i++) {
        GtkListBoxRow *row = create_row(&win->entries[i], &win->config);
        gtk_list_box_append(win->listbox, GTK_WIDGET(row));
    }

    GtkListBoxRow *first = gtk_list_box_get_row_at_index(win->listbox, 0);
    if (first) {
        gtk_list_box_select_row(win->listbox, first);
        gtk_widget_grab_focus(GTK_WIDGET(first));
    }
}

/* ── Public ────────────────────────────────────────────────────── */

PickerWindow *picker_window_new(GtkApplication *app,
                                ClipboardEntry *entries, int n_entries,
                                const Config *config) {
    PickerWindow *win = g_new0(PickerWindow, 1);
    win->app       = app;
    win->config    = *config;
    win->entries   = entries;
    win->n_entries = n_entries;

    g_mutex_init(&win->cache_mutex);
    win->image_cache = g_hash_table_new_full(
        g_str_hash, g_str_equal, g_free, image_data_free);

    /* Create window */
    win->window = GTK_WINDOW(gtk_window_new());
    gtk_window_set_application(win->window, app);

    /* Layer Shell setup */
    gtk_layer_init_for_window(win->window);
    gtk_layer_set_layer(win->window, GTK_LAYER_SHELL_LAYER_OVERLAY);
    gtk_layer_set_keyboard_mode(win->window,
        GTK_LAYER_SHELL_KEYBOARD_MODE_EXCLUSIVE);
    gtk_layer_set_namespace(win->window, "cliphist-picker");
    gtk_layer_set_exclusive_zone(win->window, -1);

    gtk_layer_set_anchor(win->window, GTK_LAYER_SHELL_EDGE_TOP, TRUE);
    gtk_layer_set_anchor(win->window, GTK_LAYER_SHELL_EDGE_BOTTOM, TRUE);
    gtk_layer_set_anchor(win->window, GTK_LAYER_SHELL_EDGE_LEFT, TRUE);
    gtk_layer_set_anchor(win->window, GTK_LAYER_SHELL_EDGE_RIGHT, TRUE);

    gtk_widget_remove_css_class(GTK_WIDGET(win->window), "background");
    gtk_widget_add_css_class(GTK_WIDGET(win->window), "backdrop");

    build_ui(win);
    setup_input(win);
    populate(win);
    load_thumbnails_async(win);

    g_signal_connect(win->window, "realize", G_CALLBACK(on_realize), win);

    return win;
}
